SSA/Projects/Scf/Scf.lean

```diff
@@ -223,16 +291,68 @@ abbrev RegionRet [Goedel Ty] (t : Ty) {Γ : Ctxt Ty} (v : Var Γ t) : Com Op Γ
   just fetching the loop variable. -/
 theorem for_return {t : Ty} (istart istep: Var Γ .int) (niters : Var Γ .nat) (v : Var Γ t) :
   Expr.denote (ScfRegion.for_ (t := t) istart istep niters v (RegionRet t ⟨1, by simp⟩)) Γv = Γv v := by
-    simp[Expr.denote, run, for_]
+    simp [Expr.denote, run, for_]
     simp_peephole at Γv
-    simp[Com.denote]
-    rw[loop_counter_decorator_constant_iterate]
+    simp [Com.denote]
+    rw [loop_counter_decorator_constant_iterate]
 -- TODO: make a `PeepholeRewrite` for `for_return`.
 
 
 set_option pp.proofs false
 set_option pp.proofs.withType false
 
+/-# Helpers for simplifying context manipulation with toSnoc and variable access -/
+section ValuationVariableAccess
+
+/-- (ctx.snoc v₁) ⟨1, _⟩ = ctx ⟨0, _⟩ -/
+@[simp]
+theorem Ctxt.Valuation.snoc_eval_one {ty : Ty} (Γ : Ctxt Ty) (V : Γ.Valuation) (v : ⟦ty⟧)
+  (hvar : Ctxt.get? (Ctxt.snoc Γ ty) 1 = some var_val) :
+  (V.snoc v) ⟨1, hvar⟩ = V ⟨0, by simp [Ctxt.get?,Ctxt.snoc] at hvar; exact hvar⟩ := rfl
+
+/-- (ctx.snoc v₁) ⟨2, _⟩ = ctx ⟨1, _⟩ -/
+@[simp]
+theorem Ctxt.Valuation.snoc_eval_two {ty : Ty} (Γ : Ctxt Ty) (V : Γ.Valuation) (v : ⟦ty⟧)
+  (hvar : Ctxt.get? (Ctxt.snoc Γ ty) 2 = some var_val) :
+  (V.snoc v) ⟨2, hvar⟩ = V ⟨1, by simp [Ctxt.get?,Ctxt.snoc] at hvar; exact hvar⟩ := rfl
+
+/-- (ctx.snoc v₁) ⟨3, _⟩ = ctx ⟨2, _⟩ -/
+@[simp]
+theorem Ctxt.Valuation.snoc_eval_three {ty : Ty} (Γ : Ctxt Ty) (V : Γ.Valuation) (v : ⟦ty⟧)
+  (hvar : Ctxt.get? (Ctxt.snoc Γ ty) 3 = some var_val) :
+  (V.snoc v) ⟨3, hvar⟩ = V ⟨2, by simp [Ctxt.get?,Ctxt.snoc] at hvar; exact hvar⟩ := rfl
+
+/-- (ctx.snoc v₁) ⟨4, _⟩ = ctx ⟨3, _⟩ -/
```

Comment:
It feels weird that these statements are hardcoded for speficic constants, can we make this a general theorem about `(V.snoc v) ⟨n+1, hvar⟩ = V ⟨n, by simp [Ctxt.get?,Ctxt.snoc] at hvar; exact hvar⟩`?